sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+
+        tmp = zero
+
+        for feature_idx in range(n_features):
+            center_diff = (
+                previous_center[feature_idx, sample_idx]
+                - center[feature_idx, sample_idx]
+            )
+            tmp += center_diff * center_diff
+
+        center_shift[sample_idx] = tmp
+
+    return center_shift[global_size, work_group_size]
+
+
+@lru_cache
+def make_half_l2_norm_dim0_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def half_l2_norm(X, result):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        l2_norm = zero
+
+        for feature_idx in range(n_features):
+            item = X[feature_idx, sample_idx]
+            l2_norm += item * item
+
+        result[sample_idx] = l2_norm / 2
+
+    return half_l2_norm[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_1dim_kernel(n_samples, work_group_size, device, dtype):
+    local_n_iterations = math.floor(math.log2(work_group_size))
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group_id = dpex.get_group_id(0)
+        local_work_id = dpex.get_local_id(0)
+        first_work_id = local_work_id == 0
+
+        n_samples = v.shape[0]
+
+        shm = dpex.local.array(work_group_size, dtype=dtype)
+
+        first_sample_idx = group_id * work_group_size * 2
+        augend_idx = first_sample_idx + local_work_id
+        addend_idx = first_sample_idx + work_group_size + local_work_id
+
+        if augend_idx >= n_samples:
+            shm[local_work_id] = zero
+        elif addend_idx >= n_samples:
+            shm[local_work_id] = v[augend_idx]
+        else:
+            shm[local_work_id] = v[augend_idx] + v[addend_idx]
+
+        dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+        current_size = work_group_size
+        for i in range(local_n_iterations):
+            current_size = current_size // 2
+            if local_work_id < current_size:
+                shm[local_work_id] += shm[local_work_id + current_size]
+
+            dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+        if first_work_id:
+            w[group_id] = shm[0]
+
+    _steps_data = []
+    n_groups = n_samples
+    # TODO: at some point, the cost of scheduling the kernel is more than the cost of
+    # running the reduction iteration. At this point the loop should stop and then a
+    # single work item should iterates one time on the remaining values to finish the
+    # reduction.
+    while n_groups > 1:
+        n_groups = math.ceil(n_groups / (2 * work_group_size))
+        global_size = n_groups * work_group_size
+        _steps_data.append(
+            (
+                sum_reduction_kernel[global_size, work_group_size],
+                dpctl.tensor.empty(n_groups, dtype=dtype, device=device),
+            )
+        )
```

Comment:
I think this is more explicit and reads better.
```suggestion
        kernel = sum_reduction_kernel[global_size, work_group_size]
        empty_tensor = dpctl.tensor.empty(n_groups, dtype=dtype, device=device)
        kernels_and_empty_tensors_pairs.append((kernel, empty_tensor))
```